tools/mkcomposefs.c

```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
 	return fwrite(buf, 1, count, file);
 }
 
+static void log_message_with_time( const char* msg )
+{
+	time_t t 		= time(NULL);
+  	struct tm tm 	= *localtime(&t);
+	fprintf(stdout, "%02d:%02d:%02d\t %s\n", 
+				tm.tm_hour, tm.tm_min, tm.tm_sec, msg);
+}
+
+struct ParallelCopyInfo
+{
+	struct lcfs_node_s		*node;
+	struct ParallelCopyInfo	*next;
+	char 					*parentPath;
+};
+
+static struct ParallelCopyInfo* new_parallel_copy_info(void)
+{
+	struct ParallelCopyInfo *ret = 
+		calloc(1, sizeof( struct ParallelCopyInfo));
+	ret->next 		= NULL;
+	ret->node 		= NULL;
+	ret->parentPath = NULL;
+
+	return ret;
+}
+
+static void cleanup_copy_node(struct ParallelCopyInfo *node)
+{
+	if(node) 
+	{
+		cleanup_copy_node(node->next);
+		free(node->parentPath);
+		free(node);
+	}	
+}
+
+static struct ParallelCopyInfo* construct_copy_data( struct lcfs_node_s* node, 
```

Comment:
I don't like how this uses a current argument which is returned changed. This will just cause problems when you want to return an error or something like that. Instead create a state structure for the parallel work and put the current pointer in it, and then pass this state to all functions and modify it, then use the return value for error reporting.